File: app/src/main/java/com/example/pam_project/database/categories/CategoryDao.java
package com.example.pam_project.database.categories;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Transaction;
import androidx.room.Update;

import com.example.pam_project.database.relationships.CategoriesWithLists;

import java.util.List;

import io.reactivex.Flowable;

@Dao
public interface CategoryDao {
    @Query("SELECT * FROM Categories")
    Flowable<List<CategoryEntity>> getAllCategories();

    @Query("SELECT * FROM Categories WHERE id =:id")
    Flowable<CategoryEntity> getCategoryById(long id);

    @Insert
    void insertCategory(final CategoryEntity category);

    @Update
    void updateCategory(final CategoryEntity category);

    @Delete
    void deleteCategory(final CategoryEntity category);

    @Transaction
    @Query("SELECT * FROM Categories")
    Flowable<List<CategoriesWithLists>> getAllCategoriesWithLists();
}

File: app/src/main/java/com/example/pam_project/database/categories/CategoryDao.kt
package com.example.pam_project.database.categories

import androidx.room.*
import com.example.pam_project.database.relationships.CategoriesWithLists
import io.reactivex.Flowable

@Dao
interface CategoryDao {
    @get:Query("SELECT * FROM Categories")
    val allCategories: Flowable<List<CategoryEntity?>?>

    @Query("SELECT * FROM Categories WHERE id =:id")
    fun getCategoryById(id: Long): Flowable<CategoryEntity?>

    @Insert
    fun insertCategory(category: CategoryEntity?)

    @Update
    fun updateCategory(category: CategoryEntity?)

    @Delete
    fun deleteCategory(category: CategoryEntity?)

    @get:Query("SELECT * FROM Categories")
    @get:Transaction
    val allCategoriesWithLists: Flowable<List<CategoriesWithLists?>?>
}